File: ios/Sources/NativeAudioPlugin/Logger.swift
import Foundation
import os.log

public class Logger {
    private var osLogger: OSLog?

    public static var debugModeEnabled = false

    public init(logTag: String) {
        osLogger = OSLog(subsystem: Bundle.main.bundleIdentifier ?? "com.capgo.native-audio", category: logTag)
    }

    func error(_ message: String, _ args: CVarArg...) {
        osLog(message, args, level: .error)
    }

    func warning(_ message: String, _ args: CVarArg...) {
        osLog(message, args, level: .fault)
    }

    func info(_ message: String, _ args: CVarArg...) {
        osLog(message, args, level: .info)
    }

    func debug(_ message: String, _ args: CVarArg...) {
        osLog(message, args, level: .debug)
    }

    func verbose(_ message: String, _ args: CVarArg...) {
        osLog(message, args, level: .default)
    }

    private func osLog(_ message: String, _ args: [CVarArg], level: OSLogType = .default) {
        if !Logger.debugModeEnabled {
            return
        }
        let formatted = String(format: message, arguments: args)
        guard let osLogger else {
            return
        }
        os_log("%{public}@", log: osLogger, type: level, formatted)
    }
}

File: android/src/main/java/ee/forgr/audio/Logger.java
package ee.forgr.audio;

import android.util.Log;
import androidx.media3.common.util.UnstableApi;

@UnstableApi
public class Logger {

    private String logTag;

    // constructor
    public Logger(String logTag) {
        this.logTag = logTag;
    }

    public void setLogTag(String logTag) {
        this.logTag = logTag;
    }

    public void error(String message) {
        if (NativeAudio.debugEnabled) {
            Log.e(logTag, message);
        }
    }

    public void error(String message, Throwable throwable) {
        if (NativeAudio.debugEnabled) {
            Log.e(logTag, message, throwable);
        }
    }

    public void warning(String message) {
        if (NativeAudio.debugEnabled) {
            Log.w(logTag, message);
        }
    }

    public void info(String message) {
        if (NativeAudio.debugEnabled) {
            Log.i(logTag, message);
        }
    }

    public void debug(String message) {
        if (NativeAudio.debugEnabled) {
            Log.d(logTag, message);
        }
    }

    public void verbose(String message) {
        if (NativeAudio.debugEnabled) {
            Log.v(logTag, message);
        }
    }
}
